js/data_structures.js

```diff
@@ -80,3 +80,62 @@ export class PriorityQueueSet {
 		return first?.value;
 	}
 }
+
+export class MaxSizeStackSet {
+	constructor(maxSize) {
+		this.maxSize = maxSize;
+		this.clear();
+	}
+
+	/**
+	 * Add an element to the top of the stack, remove any other elements with the same
+	 * value and, if then the stack is too big, remove the bottom
+	 */
+	push(value) {
+		const newNode = {
+			value,
+			next: this.top,
+			previous: null
+		}
+
+		// Add the new node to the top
+		this.top = newNode;
+		this.last ||= newNode;
```

Comment:
If there's no strict reason why all falsy-types should be overriden, I'd prefer `??=` to only override nullish-types, which I think fits the purpose here more closely.

Reply:
This whole class is gone now. But yes, you were right!